python/cuml/feature_extraction/_tfidf.py

```diff
@@ -90,11 +102,60 @@ class TfidfTransformer:
 
     def __init__(self, *, norm='l2', use_idf=True, smooth_idf=True,
                  sublinear_tf=False):
+
+        super(TfidfTransformer, self).__init__(...)
         self.norm = norm
         self.use_idf = use_idf
         self.smooth_idf = smooth_idf
         self.sublinear_tf = sublinear_tf
 
+    @with_cupy_rmm
+    def _set_doc_stats(self, X):
+        """
+        We set the following document level statistics here:
+        n_samples
+        n_features
+        df(document frequency)
+        """
+        # Should not have a cost if allready sparse
+        output_dtype = _get_dtype(X)
+        X = self._convert_to_csr(X, output_dtype)
+        n_samples, n_features = X.shape
+        df = _sparse_document_frequency(X)
+        df = df.astype(output_dtype, copy=False)
+        # TODO Finding it difficult to
+        # map below to cumlarray
+        # removed trailing underscores for now
+        self.df = df
```

Comment:
This is a temporary array that we create.  What is the best way to mark this as a member of this class  ?

Reply:
what do you mean exactly with marking it as member of the class?

Reply:
So I want to make these a `super private` attribute (https://github.com/rapidsai/cuml/issues/2456#issuecomment-666106406)  but I run into a bunch of `CumlArray` related test failures.  

What is the best way to create a temporary array that I can just keep as a cupy array or is there a better convention i should follow ?

Reply:
Currently, I am getting a failure in tests here. 

https://github.com/rapidsai/cuml/blob/196cde0b36834849393281e54d4b9823a461a0d1/python/cuml/feature_extraction/_tfidf_vectorizer.py#L202

```python
              pass
        elif (supported_type is not None):
            # Is this an estimated property? If so, should always be CumlArray
>           assert not name.endswith("_"), "Invalid Estimated Array-Like \
                Attribute! Estimated attributes should always be CumlArray. \
                Attribute: '{}' In: {}".format(name, self.__repr__())
E           AssertionError: Invalid Estimated Array-Like                 Attribute! Estimated attributes should always be CumlArray.                 Attribute: 'idf_' In: TfidfTransformer()
E           assert not True
E            +  where True = <built-in method endswith of str object at 0x7feaab54d6f0>('_')
E            +    where <built-in method endswith of str object at 0x7feaab54d6f0> = 'idf_'.endswith

conftest.py:117: AssertionError
========================================================================= short test summary info ==========================================================================
FAILED test_text_feature_extraction.py::test_tfidf_vectorizer_idf_setter - AssertionError: Invalid Estimated Array-Like                 Attribute! Estimated attributes s...
============================================================================
```


I am not really sure how are we supposed to store and create members if we want to derive from the `Base` class ? 

Like, in this case, the setter creates a sparse diagonal array while the getter should return a cupy array and there is not really interoperability of input and output are sparse matrices.


Sorry for all the questions, kind of confused by this .

Reply:
Ok the problem is that this is a very different attribute than the majority of attributes we're checking that are cumlarray (and doesn't return the sparse array), so the check doesn't apply right now. And overall tfidf and quite a few classes in feature_extraction are quite different than the bulk of cuML, so the checks don't apply as they are. The tests (and cumlarray management) will change/improve in 0.16, so this type of scenario will be taken into account. For now I would suggest adding an exception here

https://github.com/rapidsai/cuml/blob/196cde0b36834849393281e54d4b9823a461a0d1/python/cuml/test/conftest.py#L120

essentially  adding before that line something like 

```python
if name != idf_:
```
so that this parameter is not checked at least for now. I think afterwards with the improvements we can add a flag or parameter for classes that inherit from `Base` but want more flexibility in `CumlArray` usage.

Reply:
> that inherit from Base but want more flexibility in CumlArray usage

This will be great.   

The other question I have is what is the recommended way to store temporary super-private objects when inheriting from the base ? 

Is the recommendation to just store them as cuml_array and access the underlying cupy array using `to_output(out_type)` . ?

Reply:
it depends, but in general it is perfectly fine to store it as a pure cupy array for this case as far as I can see.

Reply:
Resolved the idf issue by below: https://github.com/rapidsai/cuml/pull/2698/commits/85ac30c673444b3c8c80c0e78089aa0199413d65

Have started a more detailed discussion about this at https://github.com/rapidsai/cuml/issues/2704#issuecomment-675216038 . 

Let's continue the discussion there.